### PR TITLE
v0.5: Upgrade deps, huge refactor, swap store

### PR DESCRIPTION
**Pull request by winstonli**

### CODE REVIEW

**Comment by mmazour** [comment]
Incremental review: e9885b5 :thumbsup:

**Comment by mmazour** [comment]
Incremental review: 545336e :thumbsup:

**Comment by mmazour** [comment]
:warning: force-pushed!
- moved e9885b5 and 545336e into #18 
- merged #18 and rebased #17 on master

**Comment by winstonli** [comment]
Ouch, didn't check for that force-push! Oh well. Here are changes that:
1. remove the swap_table
2. fix some empty directories in the tests
3. add new relic support

**Comment by mmazour** [comment]
Sorry, Winston!  (That's why I put the :warning: on the PR when I do that. But I'll email you directly next time as well.)

Incremental review:  f26ed19 :thumbsup:

**Comment by mmazour** [comment]
Incremental review: 0ad01c61 :thumbsup:

**Comment by mmazour** [comment]
:warning: Moved f26ed19 and 0ad01c6 into #19, which has been merged and deployed.  This branch now needs cleanup and rebase, which I'll work on today.

**Comment by mmazour** [comment]
:warning: force-pushed!
- Rebased on master (which has f26ed19 and 0ad01c6 in it now)
- Removed "that should have been a fast-forward" commit to restore mergeability

@winstonli can you have a look and be sure everything's here that you expect?  I've checked it myself, but it would be good for you to confirm.

File: src/main/java/uk/ac/ic/wlgitbridge/server/GitBridgeServer.java [hunk 1]
```diff
@@ -50,7 +51,22 @@ public GitBridgeServer(Config config) throws ServletException, InvalidRootDirect
         this.rootGitDirectoryPath = config.getRootGitDirectory();
         RepoStore repoStore = new FSRepoStore(rootGitDirectoryPath);
         DBStore dbStore = new SqliteDBStore(repoStore.getRootDirectory());
-        SwapStore swapStore = new SwapStore() {};
+        SwapStore swapStore = new SwapStore() {
+            @Override
+            public void upload(String projectName, InputStream uploadStream, long contentLength) {
```
Comment (mmazour) [inline src/main/java/uk/ac/ic/wlgitbridge/server/GitBridgeServer.java]: This is something I'm not used to.  We're creating an anonymous empty implementation of the `SwapStore` interface here, right?  Is this something that's actually OK to do, or is this just a placeholder until it's replaced with a `SwapStore.fromConfig` call in a later commit?

(Pardon the question – I've been away from Java for some years, and I'm not sure this was possible in the Java I used back then.)

**Comment by mmazour** [comment]
## General comments:
- Now that there are tests, could the `README` be updated with information on running them?
- What if a swap run takes longer than `interval`?  Should `doSwap` test to be sure it's not already running?  If there are a lot of projects to evict, could `doSwap` use enough CPU or I/O to make the bridge unavailable?  Should there be throttling of some kind?
- Should we perhaps have logic to prevent `NoopSwapStore` from being used if environment is `production`?

## Incremental reviews:
1. 5f2beaf Add implementations, implement S3SwapStore (with only tars), FSRepoStore, Tar and File utils, add tests – mostly :thumbsup: but a few notes:
   - My question about the anonymous empty implementation of `SwapStore` in `GitBridgeServer.java`
   - Naming of tests isn't consistent.  FWIW I like the names that can be read as an assertion, e.g. `startingTimerAlwaysCausesASwap` more than the ones that are less descriptive, e.g.  `testPurgeNonexistentProjects` 
2. 92f00d9 Implement bz2 support and test :thumbsup:
3. bfe1239 Change DBStore to target a file :thumbsup:
4. 6ccf5b2 Remove DB fluff layer :thumbsup:
5. 50c5f75 Catch more exceptions in db init :thumbsup:
6. 2f8d395 Move stuff to bridge package :thumbsup:
7. fc5dede Implement and test DBStore swap_table functionality – :thumbsup: apart from note about "Evicated"
8. 1fac865 Add and test an SQL update for only setting a project's last_accessed if it's missing – :thumbsup: as it stands, though if we wanted to clean up the history this seems like it could go away since it's all about the swap table.
9. d782e03 Add and test InMemorySwapStore :thumbsup:
10. b4909b2 Add and test GetNumProjects – :thumbsup: as it stands, though if we're going to have so many commits (sometimes one per query), then why is are `SwapJobConfig` and `SwapJobImpl` buried in the `GetNumProjects` commit?
11. 54e5679 Implement and test SwapJobImpl – related to the above, why start implementation `SwapJobImpl` in a previous commit and wait to implement/test/rewrite the implementation in this one?  And on the flipside of the previous commit, this time we have a `SwapJobImpl` commit which has a  `GetNumUnswappedProjects` query buried in it.  Still, :thumbsup: code-wise.
12. b011fd7 Delete the SnapshotRepositoryBuilder and WLBridgedProject abominations. Set accessed time after update and push. :thumbsup:
13. 504b746 Implement and test the swap job, and add integration test :thumbsup:
14. 9596658 Sanitise s3 info and fix for s3 :thumbsup:
15. b36b2a4 Add checkDB to upgrade the DB :thumbsup:
16. 551eaaf Update example conf and README :thumbsup:
17. 573f221 Ignore .idea and *.iml again. :thumbsup:
18. 1cd2c90 README formatting :thumbsup:
19. 743c202 Remove swap table and put last_accessed column in projects –✋  :warning: why does `createTables` invoke `new CreateProjectsIndexLastAccessed()` twice?
20. 1903268 Put files in directories that weren't commited :thumbsup:
21. 2f7b643 Add check for last_accessed column in case adding failed due to an error rather than because it already exists :thumbsup:
22. bcf0ef2 Add New Relic stuff :thumbsup:

## Review of changeset as a whole

Seems good apart from my questions above.

@winstonli, can you have a look at the above questions and comments and see what you think?

File: src/test/java/uk/ac/ic/wlgitbridge/bridge/db/sqlite/SqliteDBStoreTest.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+                "older",
+                Timestamp.valueOf(
+                        LocalDateTime.now()
+                )
+        );
+        assertEquals("asdf", dbStore.getOldestUnswappedProject());
+    }
+
+    @Test
+    public void noOldestProjectIfAllEvicated() {
```
Comment (mmazour) [inline src/test/java/uk/ac/ic/wlgitbridge/bridge/db/sqlite/SqliteDBStoreTest.java]: Evicated?

File: src/main/java/uk/ac/ic/wlgitbridge/bridge/Bridge.java [hunk 3]
```diff
@@ -110,62 +114,109 @@ public void startSwapJob() {
...
-                                          String hostname)
-            throws SnapshotPostException, IOException, ForbiddenException {
-        lock.lockForProject(projectName);
-        CandidateSnapshot candidate = null;
-        try {
-            Log.info("[{}] Pushing", projectName);
-            String postbackKey = postbackManager.makeKeyForProject(projectName);
-            Log.info(
-                    "[{}] Created postback key: {}",
+    private void updateRepositoryCritical(
```
Comment (mmazour) [inline src/main/java/uk/ac/ic/wlgitbridge/bridge/Bridge.java:174]: Why are these methods named `Critical`?

File: src/main/java/uk/ac/ic/wlgitbridge/bridge/db/sqlite/SqliteDBStore.java [hunk 4]
```diff
@@ -141,12 +131,17 @@ private Connection openConnectionTo(File dbFile) {
...
+            /* We need to eat exceptions from here */
+        }
         Stream.of(
                 new CreateProjectsTableSQLUpdate(),
+                new CreateProjectsIndexLastAccessed(),
                 new CreateURLIndexStoreSQLUpdate(),
                 new CreateIndexURLIndexStore(),
-                new CreateSwapTable(),
-                new CreateSwapTableIndex()
+                new CreateProjectsIndexLastAccessed()
```
Comment (mmazour) [inline src/main/java/uk/ac/ic/wlgitbridge/bridge/db/sqlite/SqliteDBStore.java]: Why is `new CreateProjectsIndexLastAccessed()` in there twice?

File: newrelic.yml [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+  # For example, to report data to "My Application" and "My Application 2" use this:
+  # app_name: My Application;My Application 2
+  # This setting is required. Up to 3 different application names can be specified.
+  # The first application name must be unique.
+  app_name: Git Bridge
+
+  # To enable high security, set this property to true. When in high
+  # security mode, the agent will use SSL and obfuscated SQL. Additionally,
+  # request parameters and message parameters will not be sent to New Relic.
+  high_security: false
```
Comment (mmazour) [inline newrelic.yml:37]: Why not `true`?
_Edit:_ never mind.  It looks like the ssl and sql settings below this line provide enough security.

File: src/test/java/uk/ac/ic/wlgitbridge/bridge/db/sqlite/SqliteDBStoreTest.java (hunk 2 above)
Comment (winstonli) [reply src/test/java/uk/ac/ic/wlgitbridge/bridge/db/sqlite/SqliteDBStoreTest.java]: Oops, muscle memory on that key combo.

File: src/main/java/uk/ac/ic/wlgitbridge/bridge/Bridge.java (hunk 3 above)
Comment (winstonli) [reply src/main/java/uk/ac/ic/wlgitbridge/bridge/Bridge.java:174]: It's supposed to be the "critical section" of methods that acquire locks. Do you think it needs changing?

**Comment by winstonli** [comment]
Thanks so much for reviewing this. In response to your points:

**Now that there are tests, could the README be updated with information on running them?**

OK, I'll add this to the current release.

**What if a swap run takes longer than interval? Should doSwap test to be sure it's not already running?**

This is a good point. I think the best approach for correctness is to schedule the next job after the previous one finishes. I'll add it to the current release.

**If there are a lot of projects to evict, could doSwap use enough CPU or I/O to make the bridge unavailable? Should there be throttling of some kind?**

The job will definitely be CPU or I/O bound, but I think it's unlikely that the bridge will be unavailable. I think the simplest way to throttle it would be to have the low/high watermarks close together (leading to less projects swapped more frequently).

Perhaps for the next release we can opportunistically swap based on how many projects are being accessed concurrently. The Git Bridge could detect downtime (likely at night) and swap projects until the low watermark. Before doing this, we should think about the size of the working set and how many new projects are being added.

**Should we perhaps have logic to prevent NoopSwapStore from being used if environment is production?**

This doesn't hurt, but how do I tell if we are in prod? I don't currently make use of any env vars or anything.

**My question about the anonymous empty implementation of SwapStore in GitBridgeServer.java**

This will have been temporary to keep things compiling.

**Naming of tests isn't consistent. FWIW I like the names that can be read as an assertion, e.g. startingTimerAlwaysCausesASwap more than the ones that are less descriptive, e.g. testPurgeNonexistentProjects**

I've been using assertion-based names to describe behaviour (which usually use mocking) and the less descriptive ones for general testing of a method call and state checking. But you're right, it could do with more descriptive names so I'll keep this in mind.

**... as it stands, though if we're going to have so many commits (sometimes one per query), then why is are SwapJobConfig and SwapJobImpl buried in the GetNumProjects commit? ... related to the above, why start implementation SwapJobImpl in a previous commit and wait to implement/test/rewrite the implementation in this one? And on the flipside of the previous commit, this time we have a SwapJobImpl commit which has a GetNumUnswappedProjects query buried in it.**

I guess that's the order and pacing in which I did things at the time. I didn't notice that I did this, but I think it's usually to keep things compiling. I'll keep this in mind for future commits.

**why does createTables invoke new CreateProjectsIndexLastAccessed() twice?**

I think this was by accident. I'll remove it for the current release.

**Comment by winstonli** [comment]
# To implement for this release:
1. Update README about running tests.
2. Schedule next swap job after previous job finishes.
3. Remove duplicate call to create the last_accessed index.

**Comment by winstonli** [comment]
OK, it looks good to go.

**Comment by winstonli** [comment]
OK, it looks good to go.

**Comment by mmazour** [comment]
Fantastic, thanks.  I'll give it a quick read-through.

**Comment by mmazour** [comment]
Thanks, Winston!  All looks good to me.  👍  :shipit:

**Comment by winstonli** [comment]
Sorry, I managed to miss this comma (I could've sworn I'd run the tests). But yes, looks good now.

**Comment by mmazour** [comment]
:thumbsup:

**Pull request merged**